test/create.js

```diff
@@ -76,6 +77,36 @@ test('custom headers (extend)', async t => {
 	t.is(headers.unicorn, 'rainbow');
 });
 
+test('extend overwrites arrays', t => {
+	const statusCodes = [408];
+	const a = got.extend({retry: {statusCodes}});
+	t.deepEqual(a.defaults.options.retry.statusCodes, statusCodes);
+	t.is(a.defaults.options.retry.statusCodes, statusCodes);
+});
```

Comment:
Consider this example:

```js
test('extend overwrites arrays', t => {
	let statusCodes = [408];
	const a = got.extend({retry: {statusCodes}});
	statusCodes[0] = 500;
	t.deepEqual(a.defaults.options.retry.statusCodes, [408]);
	t.not(a.defaults.options.retry.statusCodes, statusCodes);
});
```

It shouldn't fail. It should deep copy arrays IMO.

Reply:
Oh. Defaults are deep frozen here: https://github.com/sindresorhus/got/blob/da7f055749780438999d459dca5455a61abab8ba/source/create.js#L45

and ain't cloned before.

Reply:
Yeah, I made it only deep copy plain object, since all other types are merged by replacement.

Reply:
> Yeah, I made it only deep copy plain object, since all other types are merged by replacement.

I don't think obvious things need an explanation, do they? You've just repeated yourself: that's what your code does, doesn't it? Please come up with a real argument. I'll illustrate if you have problems with understanding mine:

1. Someone creates an array.
2. The array is passed to assignOptions through `got.extend()`.
3. It is passed by reference (please keep that word in mind my friend).
4. It's deep frozen (got.extend() -> got.create() -> deepFreeze())
5. Someone wants to reuse the array.
6. :boom:

Reply:
I understand that concern, but I intentionally kept the implementation simple, so that the results are predictable.

I was also concerned with performance, since this method is called not only by `got.extend` and `got.create`, but also by every call to `got()`.

There are an infinite number of possible reference types that inherit from object, which may also have unexpected behavior when cloned, especially when only an object's own properties are cloned.

I reluctantly chose to recursively merge source values that are plain objects when the target value is not a plain object for this reason because normalize-arguments currently requires that the options it receives are mutable. I would rather have refactored normalize-arguments, so that it selectively builds a new object from the input, but I thought that was too large a change to introduce here.

Do you think it would be helpful to explicitly mention in the `got.extend` and `got.create` that the resulting object is frozen, so that any reference types will be frozen? We could also mention that a workaround is to use something like `_.deepClone` if users need to preserve a mutable instance of a config setting.

We could also reconsider whether freezing the options tips the balance of predicable behavior nearer or farther from the user's expectations.

Reply:
Just clarify my position - I'm not strictly inflexible on the implementation, but I want to make sure my rationale is clearly expressed before we decide to change it.

@sindresorhus do you have an opinion on this topic?

Reply:
> I was also concerned with performance, since this method is called not only by got.extend and got.create, but also by every call to got().

Then why? There are many nitpicks that could improve performance **a bit** (it's unnoticeable) like this one and you're nothing about that. Please, be serious.

> Do you think it would be helpful to explicitly mention in the got.extend and got.create that the resulting object is frozen, so that any reference types will be frozen?

No. That's dumb. All plain objects are frozen and other objects like arrays are not? That doesn't have any sense. BTW, It's OK to pass references to other types than objects and arrays since `got` doesn't use them.

Reply:
@jstewmon Nice copycat work. GL ITF.

Reply:
> Do you think it would be helpful to explicitly mention in the got.extend and got.create that the resulting object is frozen, so that any reference types will be frozen? We could also mention that a workaround is to use something like _.deepClone if users need to preserve a mutable instance of a config setting.

Yes, that would be useful.